File: homework1/warmup.py
import random
import itertools

def change (cents):
	if cents < 0:
		raise ValueError, "amount cannot be negative"
	quarters, left_over = divmod (cents, 25)
	dimes, left_over = divmod (left_over, 10)
	nickels, pennies = divmod(left_over, 5)
	return (quarters, dimes, nickels, pennies)


def strip_quotes(s):
	return s.translate(None, "\"\'")

def scramble(s): 
	return "".join(random.sample(s, len(s)))

def powers_of_two(n):
	i = 1
	while i <= n:
		yield i
		i = i*2

def prefixes(s):
	i = 0
	while i <= len(s):
		yield s[:i]
		i += 1

def interleave(x,y):
	zipped = list(itertools.chain(*zip(x,y)))
	return zipped + x[len(y):] +y[len(x):]

def stutter(x):
	return interleave(x,x)
	

File: homework2/warmup.js
var change =  function (amount) {
	if (amount < 0) {
		throw "amount cannot be negative";
	}
	quarters = Math.floor(amount / 25);
	left_over = amount % 25;
	dimes = Math.floor(left_over / 10);
	left_over %= 10;
	nickels = Math.floor(left_over / 5);
	pennies = left_over % 5;
	return [quarters, dimes, nickels, pennies];
};

var stripQuotes = function (s) {
	return s.replace(/["']/g,"");
};


//http://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle

var scramble = function (s) {
    var a = s.split(""),
        n = a.length;

    for(var i = n - 1; i > 0; i--) {
        var j = Math.floor(Math.random() * (i + 1));
        var tmp = a[i];
        a[i] = a[j];
        a[j] = tmp;
    }
    return a.join("");
};

var powersOfTwo = function (limit, callback) {
	for (var x = 1; x <= limit; x = x*2) {
		callback(x);
	}
};

var prefixes = function (s, callback) {
	for (var i = 0; i <= s.length; i++) {
        var result = s.substring(0,i);
		callback(result);
	}
};

var interleave = function (x, y) {
    var max = Math.max(x.length, y.length), result = [];
    for (var i = 0; i < max; i ++) {
        if (i < x.length) result.push(x[i]);
        if (i < y.length) result.push(y[i]);
    }
    return result;
};

var stutter = function(x) {
	return interleave(x,x);
};